File: orchestrator/engine.py
"""Orchestrator context and engine protocol."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class ToolRunner(Protocol):
    """Protocol for something that can execute tool calls."""

    async def run(self, tool_name: str, params: dict) -> str:
        ...

    def get_tool_schemas(self) -> list[dict]:
        ...


@dataclass
class OrchestratorContext:
    """Mutable context passed through the orchestrator run."""

    history: list[dict] = field(default_factory=list)
    memory: dict[str, Any] = field(default_factory=dict)
    tool_runner: Any = None
    llm_client: Any = None
    model: str = ""


@runtime_checkable
class OrchestratorEngine(Protocol):
    """Protocol that all orchestrator implementations must satisfy."""

    async def run(self, user_input: str, context: OrchestratorContext) -> str:
        ...

File: src/orchestrator/engine.ts
import type { Message, StreamEvent } from '../types.js'
import type { LLMClient } from '../infra/llm-client.js'

/**
 * 最小 HookRegistry 接口 — hooks/registry.ts 将在 milestone-12 中实现完整版本
 * 这里只定义 orchestrator 所需的最小依赖
 */
export interface HookRegistry {
  fire(name: string, data: Record<string, unknown>): Promise<void>
}

/** 工具执行器接口 — capability 层将实现完整版本 */
export interface ToolRunner {
  run(toolName: string, params: Record<string, unknown>): Promise<string>
  getToolSchemas(): Record<string, unknown>[]
}

/** Orchestrator 运行时上下文 */
export type OrchestratorContext = {
  history: Message[]
  memory: Record<string, unknown>
  toolRunner: ToolRunner | null
  llmClient: LLMClient | null
  model: string
  hooks: HookRegistry | null
}

/** Orchestrator 引擎协议 — orchestrator 层的核心接口 */
export interface OrchestratorEngine {
  run(userInput: string, context: OrchestratorContext): Promise<string>
  runStream(userInput: string, context: OrchestratorContext): AsyncGenerator<StreamEvent>
}
